DB: fix how operation and collection names are recorded for complex queries (opt-in `db.query.text` on metrics, new `db.query.summary` recommended attribute) (#1482)
open-telemetry/semantic-conventions: docs/database/database-spans.md

e.g. from `IN (?, ?, ?, ?)` to `IN (?)`, as this can help with extremely long IN-clauses,
and can help control cardinality for users who choose to (optionally) add `db.query.text` to their metric attributes.

> ## Synthetic query text

Author: TODO: use it on spans

The `db.query.synthetic` attribute captures a shortened representation of a query text
which SHOULD have low-cardinality and SHOULD NOT contain any dynamic or sensitive data.

It SHOULD only be reported if instrumentation parses query text for the database call.

When query text is parsed, the parser SHOULD extract a list of operations and corresponding
collection names performed in this query.

The `db.query.synthetic` attribute value SHOULD contain a semicolon separated list
of operation and corresponding collection names formatted in the following way:

```
{operation1} {collection1}; {operation2} {collection2}; ...

> ````

Reviewer: wdyt of defining `db.query.synthetic` as a super sanitized version of `db.query.text` where the only things that are kept are operation names and collection names?

so that

> SELECT * from TABLE1, TABLE2 where ...

would become

> SELECT TABLE1 TABLE2

Reviewer 2: Using empty space as a separator may cause problems, as databases may allow table names to have empty spaces.

For instance, in MySQL, this statement is allowed:

```sql
CREATE TABLE `USER TABLE` (
    UserID INT
);
```

For a query like this

```sql
SELECT * from `USER TABLE`, A;
```

would become

```sql
SELECT USER TABLE A
```

Which seems confusing, as we might think it queries three tables.

---

If we use a comma as a separator, we got 

```sql
SELECT USER TABLE, A
```

This seems clear.

Author: It's a good point! Though I don't think anyone could make much sense out of this attribute alone or parse it in any meaningful way.

It's just a grouping key that is also sometimes readable for human beings.  I expect that the main use case for it is to get top N longest queries and then for each of them look into `db.query.text` to see what was this query about (via span attributes or metric exemplars). 

So I'm not sure if we need to worry about spaces in the table names on this attribute.

Reviewer 2: > Though I don't think anyone could make much sense out of this attribute alone or parse it in any meaningful way.

It sounds like it is a hashed-alike string with a human-readable context. There may be a use case to see the duration among different `db.query.synthetic`, so we can quickly find the slowest `db.query.synthetic` (a pattern) from a table perspective instead of a single query perspective.

This would require us to determine the exact database tables from `db.query.synthetic`. With `SELECT USER TABLE A`. We might think there are three tables: `USER`, `TABLE, ' and `A`. Or, two tables: `USER`, `TABLE A`, etc.

I know using empty spaces should be fine in most cases, but I think we should keep our design more tolerant of rare cases. Just like what we did for `db.collection.name`.

Reviewer 2: Another option: Users can add `"` to the database table for this special case. So, it can look like

```sql
SELECT "USER TABLE" "A"
```

wdyt? Would it sound too complicated?

Reviewer: would it work if we only quote tables that have spaces?

Author: what would we expect the `db.collection.name` be in the case of `` `USER TABLE` `` ? I'd expect it to contain all the characters that were provided in the query including quotes, double-quotes, backticks, non-ascii chars. etc.

That aligns with the language we have for collection name:
> to capture the value as provided by the application without attempting to do any case normalization

Then it should be represented in the same way in the `db.query.summary` - I can add some wording that targets there should follow the same as-is principle as on the attributes

> For example, a

Author: oops

> [!NOTE]
> Operations may be performed on an anonymous collection or on multiple collections
> at the same time, so there could be 0 or more collection names associated with one
> operation name.

If operation is performed on an anonymous collection, the corresponding collection
name SHOULD NOT be captured. If operation is performed on multiple collections, collection names
for this operation should be recorded as comma-separated list.

**Examples**:

- Query that performs single operation:

   ```sql
   SELECT *
   FROM   wuser_table
   WHERE  username = ?
   ```

   the corresponding `db.query.synthetic` is `SELECT wuser_table`.

- Query that performs multiple operations:

   ```sql
   INSERT INTO shipping_details
               (order_id,
               address)
   SELECT order_id,
         address
   FROM   orders
   WHERE  order_id = ?
   ```

>    the corresponding `db.query.synthetic` is `INSERT shipping_details; SELECT orders`.

Reviewer: just seeing what it would look like elsewhere:

> INSERT shipping_details SELECT orders

- Query that performs an operation that's applied to multiple collections:

   ```sql
   SELECT *
   FROM   songs,
         artists
   WHERE  songs.artist_id == artists.id
   ```

>   the corresponding `db.query.synthetic` is `SELECT songs,artists`.

Reviewer: and

> SELECT songs artists

- Query that performs an operation on an anonymous table:
   ```sql
   SELECT order_date
   FROM   (SELECT *
           FROM   orders o
                  JOIN customers c
                    ON o.customer_id = c.customer_id)
   ```

>   the corresponding `db.query.synthetic` is `SELECT; SELECT orders; JOIN orders,customers`.

Reviewer: and

> SELECT orders customers

Author: what about the first `SELECT` ? or should it be `SELECT SELECT orders customers` ? 

also it seems you dropped JOIN - I wanted to clarify that you don't consider it to be an operation, correct?

Reviewer: oh, yes, I didn't mean to skip the extra `SELECT` (though potentially we could consider removing consecutive duplicates...)

and yeah, I wasn't considering `JOIN` to be an operation, though it could be...

Author: I'd rather not remove duplicates to preserve a bit more of the original query. I added some wording to explain that things like SQL statements is a good start and clauses can be added on case-by-case basis, but there are tradeoffs. PTAL

e.g. from `IN (?, ?, ?, ?)` to `IN (?)`, as this can help with extremely long IN-clauses,
and can help control cardinality for users who choose to (optionally) add `db.query.text` to their metric attributes.

## Generating a summary of the query text

The `db.query.summary` attribute captures a shortened representation of a query text
which SHOULD have low-cardinality and SHOULD NOT contain any dynamic or sensitive data.

> [!NOTE]

> > The `db.query.text` is intended to identify individual queries. Even though

Reviewer: maybe
* The `db.query.text` attribute is ...
* The `db.query.text` value is ...
* `db.query.text` is ...

> is it sanitized if captured by default, it still could have high cardinality and
> might reach hundreds of lines of code.
>

> > The `db.query.summary` is intended to provide less granular grouping key that

Reviewer: ```suggestion
> The `db.query.summary` is intended to provide a less granular grouping key that
```

> > can be used as a span name or a metric attribute in a common case. It SHOULD

Reviewer: ```suggestion
> can be used as a span name or a metric attribute in common cases. It SHOULD
```

> only contain information that has a significant impact on the query, database,
> or application performance.

Instrumentations that support query parsing SHOULD generate query summary when
one is not readily available from other sources.

The summary SHOULD preserve the following parts of query in the order they were provided:

> - operations such as SQL SELECT, INSERT, UPDATE, DELETE or other statements

Reviewer: is "statement" the right word here?

Author: yeah, not right. I changed to "command", but it's not a precise referenceable technical term either - I don't think there is a precise term defined in the language :( 

The problem is that we want a small subset of clauses - only the top-level ones. Open to any suggestions on how we can describe it better.
The only alternative I can think of is literally document all the commands that we'd support for SQL.

E.g. here's what ChatGPT  classifies as 'commands'

- SELECT
- INSERT
- UPDATE
- DELETE
- MERGE
- CREATE
- ALTER
- DROP
- TRUNCATE
- RENAME
- GRANT
- REVOKE
- COMMIT
- ROLLBACK
- SAVEPOINT
- SET TRANSACTION
- CREATE INDEX
- DROP INDEX
- ALTER INDEX
- START TRANSACTION
- SET

which seems to align with what we want it to be, so just saying "command" might be good enough.

Reviewer: > "command" might be good enough

👍

- operation targets such as collection or database name

Instrumentation MAY include additional details such as specific SQL clauses as long
as summary remains relatively short and its cardinality remains relatively low
comparing to the `db.query.text`.

The instrumentation SHOULD parse the query and extract a list of operations and
corresponding targets from the query. It SHOULD set `db.query.summary` attribute
to the value formatted in the following way:

```
{operation1} {target1} {operation2} {target2} {target3} ...
````

**Examples**:

- Query that consist of a single operation:

   ```sql
   SELECT *
   FROM   wuser_table
   WHERE  username = ?
   ```

   the corresponding `db.query.summary` is `SELECT wuser_table`.

- Query that performs multiple operations:

   ```sql
   INSERT INTO shipping_details
               (order_id,
               address)
   SELECT order_id,

>          address

Reviewer: ```suggestion
          address
```

e.g. from `IN (?, ?, ?, ?)` to `IN (?)`, as this can help with extremely long IN-clauses,
and can help control cardinality for users who choose to (optionally) add `db.query.text` to their metric attributes.

## Generating a summary of the query text

The `db.query.summary` attribute captures a shortened representation of a query text
which SHOULD have low-cardinality and SHOULD NOT contain any dynamic or sensitive data.

> [!NOTE]
> The `db.query.text` attribute is intended to identify individual queries. Even though
> it is sanitized if captured by default, it could still have high cardinality and
> might reach hundreds of lines.
>
> The `db.query.summary` is intended to provide a less granular grouping key that

> > can be used as a span name or a metric attribute in common cases. It SHOULD
> > only contain information that has a significant impact on the query, database,
> > or application performance.

Reviewer: i'm not sure this part helps

I think I understand, that it groups things where queries in each bucket will have somewhat similar performance characteristics to each other, but worried that people would use this language to pull in more stuff into the summary than we want (increasing cardinality and reducing brevity), since there are many things that can have a significant impact (e.g. where clauses)

```suggestion
> can be used as a span name or a metric attribute in common cases.
```

Instrumentations that support query parsing SHOULD generate a query summary when
one is not readily available from other sources.

> The summary SHOULD preserve the following parts of query in the order they were provided:

Reviewer: ```suggestion
The summary SHOULD preserve the following parts of a query in the order they were provided:
```

   ```sql
   INSERT INTO shipping_details
               (order_id,
               address)
   SELECT order_id,
          address
   FROM   orders
   WHERE  order_id = ?
   ```

   the corresponding `db.query.summary` is `INSERT shipping_details SELECT orders`.

- Query that performs an operation that's applied to multiple collections:

   ```sql
   SELECT *
   FROM   songs,
          artists
   WHERE  songs.artist_id == artists.id
   ```

   the corresponding `db.query.summary` is `SELECT songs artists`.

- Query that performs an operation on an anonymous table:

   ```sql
   SELECT order_date
   FROM   (SELECT *
           FROM   orders o
                  JOIN customers c
                    ON o.customer_id = c.customer_id)
   ```

   the corresponding `db.query.summary` is `SELECT SELECT orders customers`.

- Query that performs an operation on multiple collections with double-quotes or other punctuation:

    ```sql
    SELECT *
    FROM   "song list",
           'artists'
    ```

    the corresponding `db.query.summary` is `SELECT "songs list" 'artists'`.

Semantic conventions for individual database systems or specialized instrumentations

> MAY specify a different `db.query.summary` format as long as produced summary remains
> relatively short and its cardinality remains low comparing to the `db.query.text`.

Reviewer: ```suggestion
MAY specify a different `db.query.summary` format as long as the produced summary remains
relatively short and its cardinality remains low compared to `db.query.text`.
```

e.g. from `IN (?, ?, ?, ?)` to `IN (?)`, as this can help with extremely long IN-clauses,
and can help control cardinality for users who choose to (optionally) add `db.query.text` to their metric attributes.

## Generating a summary of the query text

The `db.query.summary` attribute captures a shortened representation of a query text
which SHOULD have low-cardinality and SHOULD NOT contain any dynamic or sensitive data.

> [!NOTE]
> The `db.query.text` attribute is intended to identify individual queries. Even though
> it is sanitized if captured by default, it could still have high cardinality and

> > might reach hundreds of lines of code.

Reviewer: ```suggestion
> might reach hundreds of lines.
```

>
> The `db.query.summary` is intended to provide a less granular grouping key that
> can be used as a span name or a metric attribute in common cases. It SHOULD
> only contain information that has a significant impact on the query, database,
> or application performance.

> Instrumentations that support query parsing SHOULD generate query summary when

Reviewer: ```suggestion
Instrumentations that support query parsing SHOULD generate a query summary when
```

one is not readily available from other sources.

The summary SHOULD preserve the following parts of query in the order they were provided:

> - operations such as SQL SELECT, INSERT, UPDATE, DELETE or other commands
> - operation targets such as collection or database names

Reviewer: ```suggestion
- operations such as SQL SELECT, INSERT, UPDATE, DELETE, and other commands
- operation targets such as collections and database names
```

Instrumentation MAY include additional details such as specific SQL clauses as long
as summary remains relatively short and its cardinality remains low comparing to

> the `db.query.text`.

Reviewer: any particular SQL clauses you have in mind?

Author: I had `JOIN` in mind, but nothing else. I'm considering to remove this sentence - it doesn't seem to convey anything new.

Author: moved to the end of the section:

> Semantic conventions for individual database systems or specialized instrumentations
MAY specify a different `db.query.summary` format as long as produced summary remains
relatively short and its cardinality remains low comparing to the `db.query.text`.

> The instrumentation SHOULD parse the query and extract a list of operations and

Reviewer: ```suggestion
Instrumentations that support query parsing SHOULD parse the query and extract a list of operations and
```

Reviewer: did you mean to mark this resolved?

> The instrumentation SHOULD parse the query and extract a list of operations and
> corresponding targets from the query. It SHOULD set `db.query.summary` attribute

Reviewer: also, should we qualify "if instrumentation parses" here?

```suggestion
The instrumentation SHOULD parse the query and extract a list of operations and
targets from the query. It SHOULD set `db.query.summary` attribute
```

to the value formatted in the following way:

```
{operation1} {target1} {operation2} {target2} {target3} ...
````

Instrumentations SHOULD capture the values of operations and targets as provided
by the application without attempting to do any case normalization. If the operation
and target value is populated on `db.operation.name`, `db.collection.name`, `db.namespace`

> or other attributes, it SHOULD match the value used in the `db.query.summary`.

Reviewer: I think we've defined `db.namespace` in terms of the connection now, so not sure if it would ever be extracted from the query

Author: I think we still allow instrumentations to keep track if the current connection

> A connection's currently associated database may change during its lifetime, e.g. from executing `USE <database>`.

> If instrumentation is unable to capture the connection's currently associated database on each query without triggering an additional query to be executed (e.g. `SELECT DATABASE()`),
then it is RECOMMENDED to fallback and use the database provided when the connection was established.

>  Instrumentation SHOULD document if `db.namespace` reflects the database provided when the connection was established.

So it's technically possible if they match

Author: removed `db.namespace` from this list, it's not necessary there anyway

Reviewer: I think I see...

so someone is using database `MyDB1` and then executes `USE MyDB2`:

* `db.query.text=USE MyDB`
* `db.query.summary=USE MyDB`
* `db.operation.name=USE`
* `db.collection.name` is not captured
* `db.namespace=MyDB1` (I think, b/c that was the namespace associated with the connection at the time?)

and internally if the instrumentation is tracking namespace changes, then on the next query `SELECT * FROM table5`:

* `db.query.text=SELECT * FROM table5`
* `db.query.summary=SELECT table5`
* `db.operation.name=SELECT`
* `db.collection.name=table5`
* `db.namespace=MyDB2`

Author: I think so.

Until recently we used to say that even the first span should point to the new database: https://github.com/open-telemetry/semantic-conventions/commit/25d438c3bf7234d27ba7a55e4520943e601f3cb9

> For commands that switch the keyspace, this SHOULD be set to the target keyspace (even if the command fails).

But the old version goes against of the current attribute definition.

**Examples**:

- Query that consist of a single operation:

   ```sql
   SELECT *
   FROM   wuser_table
   WHERE  username = ?
   ```

   the corresponding `db.query.summary` is `SELECT wuser_table`.

- Query that performs multiple operations:

   ```sql
   INSERT INTO shipping_details
               (order_id,
               address)
   SELECT order_id,
          address
   FROM   orders
   WHERE  order_id = ?
   ```

   the corresponding `db.query.summary` is `INSERT shipping_details SELECT orders`.

- Query that performs an operation that's applied to multiple collections:

   ```sql
   SELECT *
   FROM   songs,
          artists
   WHERE  songs.artist_id == artists.id
   ```

  the corresponding `db.query.summary` is `SELECT songs artists`.

> 

Reviewer 2: ```suggestion

- Query that performs an operation that's applied to multiple collections with double-quotes or other punctuation:

    ```sql
    SELECT *
    FROM   "song list",
           'artists'
    ```

    the corresponding `db.query.summary` is `SELECT "songs list" 'artists'`.

```

Could we include this in examples? So it could guide future readers to deal with this situation.

Related comments: https://github.com/open-telemetry/semantic-conventions/pull/1482#discussion_r1803546355

Author: added!
